### PR TITLE
[4.2] Convert the checkfiles plugin to service provider

### PR DESCRIPTION
**Pull request by laoneo**
### Summary of Changes
Converts the checkfiles plugin to service provider and adds some unit tests. This pr is based on #38060, so please test that one first.

### Testing Instructions
- Create a folder in images with the name test and upload there a big image from unsplash or so which is bigger than 1080 pixels
- Go to the task scheduler back end list
- Click on new "Image Size Check"
- Select the test image in the task parameters
- Save the task and close the form
- Run the task by clicking on "Run Test" button

### Actual result BEFORE applying this Pull Request
The image in the test folder has a width or height of 1080 pixels.

### Expected result AFTER applying this Pull Request
The image in the test folder has a width or height of 1080 pixels.

### CODE REVIEW

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    {
+		if (!is_dir(__DIR__ . '/tmp'))
+		{
+			mkdir(__DIR__ . '/tmp');
+		}
+
+		$image = imagecreate(200, 200);
+        imagecolorallocate($image, 255, 255, 0);
+        imagepng($image, __DIR__ . '/tmp/test.png');
+        imagedestroy($image);
```
Comment (richard67) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
		$image = imagecreate(200, 200);
        imagecolorallocate($image, 255, 255, 0);
        imagepng($image, __DIR__ . '/tmp/test.png');
        imagedestroy($image);
```
```suggestion
		$image = imagecreate(200, 200);
		imagecolorallocate($image, 255, 255, 0);
		imagepng($image, __DIR__ . '/tmp/test.png');
		imagedestroy($image);
```
@laoneo You really should adjust your editor so you can see what is a tab and what are spaces, and switch off any auto indentation features. I found this place here now where a mix was used. I haven't checked at other places of this PR. Please do so.

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+{
+	/**
+	 * Setup
+	 *
+	 * @return  void
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function setUp(): void
+    {
```
Comment (richard67) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
	public function setUp(): void
	{
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+		if (!is_dir(__DIR__ . '/tmp'))
+		{
+			mkdir(__DIR__ . '/tmp');
+		}
+
+		$image = imagecreate(200, 200);
+        imagecolorallocate($image, 255, 255, 0);
+        imagepng($image, __DIR__ . '/tmp/test.png');
+        imagedestroy($image);
+    }
```
Comment (richard67) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php:52]:
```suggestion
	}
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+	 *
+	 * @since   __DEPLOY_VERSION__
+	 */
+	public function tearDown(): void
+    {
+		if (is_dir(__DIR__ . '/tmp'))
+		{
+			Folder::delete(__DIR__ . '/tmp');
+		}
+    }
```
Comment (richard67) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
	public function tearDown(): void
	{
		if (is_dir(__DIR__ . '/tmp'))
		{
			Folder::delete(__DIR__ . '/tmp');
		}
	}
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 5]
```diff
@@ -0,0 +1,221 @@
+<?php
+/**
+ * @package     Joomla.UnitTest
+ * @subpackage  Extension
+ *
+ * @copyright   (C) 2022 Open Source Matters, Inc. <https://www.joomla.org>
+ * @license     GNU General Public License version 2 or later; see LICENSE.txt
+ */
+
+namespace Joomla\Tests\Unit\Plugin\Task\Requests\Extension;
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
namespace Joomla\Tests\Unit\Plugin\Task\Checkfiles\Extension;
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+ * Test class for Requests plugin
+ *
+ * @package     Joomla.UnitTest
+ * @subpackage  Requests
+ *
+ * @testdox     The Checkfiles plugin
+ *
+ * @since       __DEPLOY_VERSION__
+ */
+class RequestsPluginTest extends UnitTestCase
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
class CheckfilesPluginTest extends UnitTestCase
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+use Joomla\Component\Scheduler\Administrator\Event\ExecuteTaskEvent;
+use Joomla\Component\Scheduler\Administrator\Task\Status;
+use Joomla\Component\Scheduler\Administrator\Task\Task;
+use Joomla\Event\Dispatcher;
+use Joomla\Filesystem\Folder;
+use Joomla\Plugin\Task\Checkfiles\Extension\Checkfiles;
+use Joomla\Tests\Unit\UnitTestCase;
+
+/**
+ * Test class for Requests plugin
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
 * Test class for Checkfiles plugin
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+ *
+ * @package     Joomla.UnitTest
+ * @subpackage  Requests
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
 * @subpackage  Checkfiles
```

File: plugins/task/checkfiles/checkfiles.xml [hunk 9]
```diff
@@ -9,9 +9,10 @@
 	<authorUrl>www.joomla.org</authorUrl>
 	<version>4.1</version>
 	<description>PLG_TASK_CHECK_FILES_XML_DESCRIPTION</description>
+	<namespace path="src">Joomla\Plugin\Task\Checkfiles</namespace>
 	<files>
-		<filename plugin="checkfiles">checkfiles.php</filename>
-		<folder>language</folder>
+		<folder plugin="requests">services</folder>
```
Comment (heelc29) [inline plugins/task/checkfiles/checkfiles.xml]:
```suggestion
		<folder plugin="checkfiles">services</folder>
```

**Review by heelc29 — CHANGES_REQUESTED** [review]

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 10]
```diff
@@ -65,6 +62,36 @@ public static function getSubscribedEvents(): array
...
+	 * @since  __DEPLOY_VERSION__
+	 */
+	private $rootDirectory;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param   DispatcherInterface  $dispatcher     The dispatcher
+	 * @param   array                $config         An optional associative array of configuration settings
+	 * @param   string               $rootDirectory  The root directory to lok for images
```
Comment (heelc29) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]:
```suggestion
	 * @param   string               $rootDirectory  The root directory to look for images
```

**Comment by alikon** [comment]
I have tested this item :white_check_mark: successfully on db97381661de915288cb4d52ec6b2ed36caa2507<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/38072">issues.joomla.org/tracker/joomla-cms/38072</a>.</sub>

**Comment by HLeithner** [comment]
> A new pull request has been created automatically to convert this PR to the PSR-12 coding standard. The pr can be found at [Digital-Peak#24](https://github.com/Digital-Peak/joomla-cms/pull/24)

there was a merge error on this pull request please check if worked correctly

**Comment by laoneo** [comment]
I had to redo this pr, @heelc29 can you have a look again if I missed something. Thanks!

File: plugins/task/checkfiles/checkfiles.xml (hunk 9 above)
Comment (heelc29) [inline plugins/task/checkfiles/checkfiles.xml]:
```suggestion
		<folder plugin="checkfiles">services</folder>
```

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 11]
```diff
@@ -65,6 +60,36 @@ public static function getSubscribedEvents(): array
...
+     * @since  __DEPLOY_VERSION__
+     */
+    private $rootDirectory;
+
+    /**
+     * Constructor.
+     *
+     * @param   DispatcherInterface  $dispatcher     The dispatcher
+     * @param   array                $config         An optional associative array of configuration settings
+     * @param   string               $rootDirectory  The root directory to lok for images
```
Comment (heelc29) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]:
```suggestion
     * @param   string               $rootDirectory  The root directory to look for images
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php (hunk 5 above)
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
namespace Joomla\Tests\Unit\Plugin\Task\Checkfiles\Extension;
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php (hunk 7 above)
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
 * Test class for Checkfiles plugin
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php (hunk 8 above)
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
 * @subpackage  Checkfiles
```

**Review by heelc29 — COMMENTED** [review]
the same changes 😅

**Comment by laoneo** [comment]
I know, somehow the psr change screwed up this pr and I had to redo it from an old stage. Sorry for that and thanks for your help.

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 12]
```diff
@@ -0,0 +1,221 @@
+<?php
+/**
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
<?php

/**
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 13]
```diff
@@ -0,0 +1,221 @@
...
+     * Setup
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function setUp(): void
+    {
+        if (!is_dir(__DIR__ . '/tmp'))
+        {
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        if (!is_dir(__DIR__ . '/tmp')) {
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 14]
```diff
@@ -0,0 +1,221 @@
...
+     * Cleanup
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function tearDown(): void
+    {
+        if (is_dir(__DIR__ . '/tmp'))
+        {
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        if (is_dir(__DIR__ . '/tmp')) {
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 15]
```diff
@@ -0,0 +1,221 @@
...
+     */
+    public function testResize()
+    {
+        $language = $this->createStub(Language::class);
+        $language->method('_')->willReturn('test');
+
+        $app = $this->createStub(CMSApplicationInterface::class);
+        $app->method('getLanguage')->willReturn($language);
+
+        $plugin = new Checkfiles(new Dispatcher, [], __DIR__);
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        $plugin = new Checkfiles(new Dispatcher(), [], __DIR__);
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 16]
```diff
@@ -0,0 +1,221 @@
...
+    {
+        copy(__DIR__ . '/tmp/test.png',__DIR__ . '/tmp/test1.png');
+
+        $language = $this->createStub(Language::class);
+        $language->method('_')->willReturn('test');
+
+        $app = $this->createStub(CMSApplicationInterface::class);
+        $app->method('getLanguage')->willReturn($language);
+
+        $plugin = new Checkfiles(new Dispatcher, [], __DIR__);
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        $plugin = new Checkfiles(new Dispatcher(), [], __DIR__);
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 17]
```diff
@@ -0,0 +1,221 @@
...
+     */
+    public function testIgnoreResize()
+    {
+        $language = $this->createStub(Language::class);
+        $language->method('_')->willReturn('test');
+
+        $app = $this->createStub(CMSApplicationInterface::class);
+        $app->method('getLanguage')->willReturn($language);
+
+        $plugin = new Checkfiles(new Dispatcher, [], __DIR__);
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        $plugin = new Checkfiles(new Dispatcher(), [], __DIR__);
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 18]
```diff
@@ -0,0 +1,221 @@
...
+     */
+    public function testInvalidFolder()
+    {
+        $language = $this->createStub(Language::class);
+        $language->method('_')->willReturn('test');
+
+        $app = $this->createStub(CMSApplicationInterface::class);
+        $app->method('getLanguage')->willReturn($language);
+
+        $plugin = new Checkfiles(new Dispatcher, [], __DIR__);
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        $plugin = new Checkfiles(new Dispatcher(), [], __DIR__);
```

File: tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php [hunk 19]
```diff
@@ -0,0 +1,221 @@
...
+    /**
+     * @testdox  can resize a subset of images
+     *
+     * @return  void
+     *
+     * @since   __DEPLOY_VERSION__
+     */
+    public function testResizeWithLimit()
+    {
+        copy(__DIR__ . '/tmp/test.png',__DIR__ . '/tmp/test1.png');
```
Comment (heelc29) [inline tests/Unit/Plugin/Task/Checkfiles/Extension/CheckfilesPluginTest.php]:
```suggestion
        copy(__DIR__ . '/tmp/test.png', __DIR__ . '/tmp/test1.png');
```

**Review by heelc29 — COMMENTED** [review]
several cs errors
https://ci.joomla.org/joomla/joomla-cms/56095/1/6

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 20]
```diff
@@ -105,23 +127,22 @@ protected function checkImages(ExecuteTaskEvent $event): int
             $newHeight = $dimension === 'height' ? $limit : $height * $limit / $width;
             $newWidth  = $dimension === 'width' ? $limit : $width * $limit / $height;
 
-            $this->logTask(Text::sprintf('PLG_TASK_CHECK_FILES_LOG_RESIZING_IMAGE', $width, $height, $newWidth, $newHeight, $imageFilename));
+            $this->logTask($this->translate('PLG_TASK_CHECK_FILES_LOG_RESIZING_IMAGE', $width, $height, $newWidth, $newHeight, $imageFilename));
```
Comment (roland-d) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]: @laoneo This should be reverted as the `translate()` function no longer exists.

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 21]
```diff
@@ -76,22 +101,19 @@ public static function getSubscribedEvents(): array
...
+        $params    = $event->getArgument('params');
+        $path      = Path::check($this->rootDirectory . $params->path);
         $dimension = $params->dimension;
         $limit     = $params->limit;
         $numImages = max(1, (int) $params->numImages ?? 1);
 
-        if (!Folder::exists($path)) {
-            $this->logTask(Text::_('PLG_TASK_CHECK_FILES_LOG_IMAGE_PATH_NA'), 'warning');
+        if (!is_dir($path)) {
+            $this->logTask($this->translate('PLG_TASK_CHECK_FILES_LOG_IMAGE_PATH_NA'), 'warning');
```
Comment (roland-d) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]: @laoneo This should be reverted as the translate() function no longer exists.

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 22]
```diff
@@ -105,23 +127,22 @@ protected function checkImages(ExecuteTaskEvent $event): int
...
 
             $image = new Image($imageFilename);
 
             try {
                 $image->resize($newWidth, $newHeight, false);
             } catch (LogicException $e) {
-                $this->logTask('PLG_TASK_CHECK_FILES_LOG_RESIZE_FAIL', 'error');
-                $resizeFail = true;
-            }
+                $this->logTask($this->translate('PLG_TASK_CHECK_FILES_LOG_RESIZE_FAIL'), 'error');
```
Comment (roland-d) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]: @laoneo This should be reverted as the translate() function no longer exists.

File: plugins/task/checkfiles/src/Extension/Checkfiles.php [hunk 23]
```diff
@@ -105,23 +127,22 @@ protected function checkImages(ExecuteTaskEvent $event): int
...
 
-            if (!empty($resizeFail)) {
                 return TaskStatus::KNOCKOUT;
             }
 
             if (!$image->toFile($imageFilename, $properties->type)) {
-                $this->logTask('PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL', 'error');
+                $this->logTask($this->translate('PLG_TASK_CHECK_FILES_LOG_IMAGE_SAVE_FAIL'), 'error');
```
Comment (roland-d) [inline plugins/task/checkfiles/src/Extension/Checkfiles.php]: @laoneo This should be reverted as the translate() function no longer exists.

**Review by roland-d — CHANGES_REQUESTED** [review]

File: plugins/task/checkfiles/src/Extension/Checkfiles.php (hunk 20 above)
Comment (laoneo) [reply plugins/task/checkfiles/src/Extension/Checkfiles.php]: done

**Review by roland-d — APPROVED** [review]

**Comment by roland-d** [comment]
Thank you

**Pull request merged**